### PR TITLE
Center windows when extracting windows to a bigger output window

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

## Description
- [x] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR

This fixes a bug in the way we extract windows for STFT when the nfft argument is bigger than the window length.
The old behavior pads on the left.
The correct behavior is to center the extracted window and pad both sides, before calculating the FFT.

This PR also introduces testing of center_windows=False in Python tests when nfft==window_length. Librosa seems to disregard the requested window_length and always extract nfft sized windows. Because of that, we can't compare with librosa for this case.

#### Additional information
- Affected modules and functionalities:

Spectrogram, ExtractWindows, FFT, STFT implementations

- Key points relevant for the review:

Changes in the kernels

### Tests
- [x] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2114

### CODE REVIEW

File: dali/kernels/signal/window/extract_windows_gpu_test.cu [hunk 1]
```diff
@@ -47,23 +47,31 @@ TEST(ExtractWindowsGPU, NonBatchedKernel) {
...
+
+    int win_data_start = (outwinlen - winlen) / 2;
+    int win_data_end = win_data_start + winlen;
+
     window::ExtractVerticalWindowsKernel<<<dim3(xblocks, yblocks), dim3(32, 32)>>>(
-      out_gpu, windows, stride, in_gpu, length, nullptr, winlen, outwinlen, center, step, reflect);
+      out_gpu, windows, stride, in_gpu, length, nullptr, outwinlen, win_data_start, winlen, center, step, reflect);
     CUDA_CALL(
-      cudaMemcpy(out.data(), out_gpu, sizeof(float)*winlen*windows, cudaMemcpyDeviceToHost));
+      cudaMemcpy(out.data(), out_gpu, sizeof(float)*outwinlen*windows, cudaMemcpyDeviceToHost));
```
Comment (jantonguirao) [inline dali/kernels/signal/window/extract_windows_gpu_test.cu:57]: bug: the order of the arguments was swapped

File: dali/kernels/signal/window/extract_windows_gpu.cuh [hunk 2]
```diff
@@ -456,10 +459,12 @@ struct ExtractVerticalWindowsImplGPU : ExtractWindowsImplGPU<Dst, Src> {
       make_span(cpu_samples, N),
       make_span(cpu_blocks, nblocks));
 
+    int in_win_start = out_win_length > args.window_length ? out_win_length - args.window_length : 0;
```
Comment (mzient) [inline dali/kernels/signal/window/extract_windows_gpu.cuh:462]:
```suggestion
    int in_win_start = out_win_length > args.window_length ? (out_win_length - args.window_length) / 2 : 0;
```

File: dali/kernels/signal/window/extract_windows_gpu.cuh [hunk 3]
```diff
@@ -639,14 +651,18 @@ struct ExtractHorizontalWindowsImplGPU : ExtractWindowsImplGPU<Dst, Src> {
...
       // calculate padding setup
       if (pad_grid.z) {
-        auto *ptr = out_ptr + args.window_length;
+        auto *ptr = out_ptr;
         for (int w = 0; w < nwindows; w += pad_block_size) {
           assert(pad_blocks < static_cast<int>(pad_grid.z));
           int cnt = std::min(pad_block_size, nwindows - w);
-          cpu_pad_blocks[pad_blocks] = {
-            ptr, cnt, static_cast<int>(sample.output_stride)
+          cpu_pad_blocks[pad_blocks++] = {
```
Comment (mzient) [inline dali/kernels/signal/window/extract_windows_gpu.cuh:658]:
```suggestion
          if (pad_left > 0)
            cpu_pad_blocks[pad_blocks++] = {
```

File: dali/kernels/signal/window/extract_windows_gpu.cuh [hunk 4]
```diff
@@ -639,14 +651,18 @@ struct ExtractHorizontalWindowsImplGPU : ExtractWindowsImplGPU<Dst, Src> {
...
+            ptr, cnt, static_cast<int>(sample.output_stride), pad_left
           };
-          pad_blocks++;
+          ptr += pad_left + args.window_length;
+          cpu_pad_blocks[pad_blocks++] = {
```
Comment (mzient) [inline dali/kernels/signal/window/extract_windows_gpu.cuh:662]:
```suggestion
          if (pad_right > 0)
            cpu_pad_blocks[pad_blocks++] = {
```

File: dali/kernels/signal/window/extract_windows_gpu.cuh [hunk 5]
```diff
@@ -237,7 +240,7 @@ __global__ void ExtractVerticalWindowsBatchedKernel(
       pos,                        // first window index
       dst, num_windows, stride,   // output
       src, length,                // input
-      window, out_win_len, in_win_len, win_center, step, reflect,  // windowing options
+      window, out_win_len, in_win_start, in_win_len, win_center, step, reflect,  // win options
```
Comment (mzient) [inline dali/kernels/signal/window/extract_windows_gpu.cuh:243]: Why have you truncated `windowing`? Let me guess... linter?

File: dali/kernels/signal/window/extract_windows_gpu.cuh [hunk 6]
```diff
@@ -639,14 +651,23 @@ struct ExtractHorizontalWindowsImplGPU : ExtractWindowsImplGPU<Dst, Src> {
...
           int cnt = std::min(pad_block_size, nwindows - w);
-          cpu_pad_blocks[pad_blocks] = {
-            ptr, cnt, static_cast<int>(sample.output_stride)
-          };
-          pad_blocks++;
+          if (pad_left > 0) {
+            cpu_pad_blocks[pad_blocks++] = {
+              ptr, cnt, static_cast<int>(sample.output_stride), pad_left
+            };
+            ptr += pad_left;
```
Comment (mzient) [inline dali/kernels/signal/window/extract_windows_gpu.cuh:662]: The pointer is used in 671 and should not change here.

File: dali/kernels/signal/window/extract_windows_gpu.cuh (hunk 5 above)
Comment (jantonguirao) [reply dali/kernels/signal/window/extract_windows_gpu.cuh:243]: yes. It seemed better than breaking the line or adding NOLINT

**Comment by jantonguirao** [comment]
!build

File: dali/test/python/test_operator_spectrogram.py [hunk 7]
```diff
@@ -114,32 +115,33 @@ def iter_setup(self):
...
     eii2 = RandomDataIterator(batch_size, shape=input_shape, dtype=np.float32)
     compare_pipelines(
         SpectrogramPipeline(device, batch_size, iter(eii1), nfft=nfft, window=None,
-                            window_length=window_length, window_step=window_step),
+                            window_length=window_length, window_step=window_step, center=center),
         SpectrogramPythonPipeline(device, batch_size, iter(eii2), window=None, nfft=nfft,
-                                  window_length=window_length, window_step=window_step),
-        batch_size=batch_size, N_iterations=5, eps=1e-04)
+                                  window_length=window_length, window_step=window_step, center=center),
+        batch_size=batch_size, N_iterations=3, eps=1e-04)
```
Comment (mzient) [inline dali/test/python/test_operator_spectrogram.py:126]: :+1:

Comment (mzient) [reply dali/test/python/test_operator_spectrogram.py:126]: Perhaps it would be even faster to run CPU and GPU here and only run the Python thingy once.

**Review by mzient — APPROVED** [review]

**Review by mzient — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Review by klecki — APPROVED** [review]
Mostly docstring issues.

**Comment by jantonguirao** [comment]
!build

**Pull request merged**